lasertag_ppo.py

```diff
@@ -0,0 +1,352 @@
+
+    Lasertag Environment:
+        * Action shape:  `n_agents` * `Discrete(5)`
+        * Observation shape: Dict('image': Box(0, 255, (`n_agents`, 3, 5, 5), uint8))
+    """
+
+    def __init__(self, n_agents, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.n_agents = n_agents
+        self.task = None
+        self.episode_return = 0
+        self.task_space = TaskSpace(spaces.MultiDiscrete(np.array([[2], [5]])))
+        self.possible_agents = np.arange(self.n_agents)
+
+    def _np_array_to_pz_dict(self, array: np.ndarray) -> dict[str : np.ndarray]:
+        """
+        Returns a dictionary containing individual observations for each agent.
+        Assumes that the batch dimension represents individual agents.
+        """
+        out = {}
+        for idx, i in enumerate(array):
+            out[str(idx)] = i
+        return out
+
+    def _singleton_to_pz_dict(self, value: bool) -> dict[str:bool]:
+        """
+        Broadcasts the `done` and `trunc` flags to dictionaries keyed by agent id.
+        """
+        return {str(idx): value for idx in range(self.n_agents)}
+
+    def reset(self) -> tuple[dict[AgentID, ObsType], dict[AgentID, dict]]:
+        """
+        Resets the environment and returns a dictionary of observations
+        keyed by agent ID.
+        """
+        obs = self.env.reset()
+        pz_obs = self._np_array_to_pz_dict(obs["image"])
+
+        return pz_obs
+
+    def step(self, action: dict[AgentID, ActionType], device: str) -> tuple[
+        dict[AgentID, ObsType],
+        dict[AgentID, float],
+        dict[AgentID, bool],
+        dict[AgentID, bool],
+        dict[AgentID, dict],
+    ]:
+        """
+        Takes inputs in the PettingZoo (PZ) Parallel API format, performs a step and
+        returns outputs in PZ format.
+        """
+        action = batchify(action, device)
+        obs, rew, done, info = self.env.step(action)
+        obs = obs["image"]
+        trunc = 0  # there is no `truncated` flag in this environment
+        self.task_completion = self._task_completion(obs, rew, done, trunc, info)
+        # convert outputs back to PZ format
+        obs, rew = tuple(map(self._np_array_to_pz_dict, [obs, rew]))
+        done, trunc, info = tuple(
+            map(self._singleton_to_pz_dict, [done, trunc, self.task_completion])
+        )
+
+        return self.observation(obs), rew, done, trunc, info
+
+
+class SelfPlay(Curriculum):
+    def __init__(self, agent, device: str, store_agents_on_cpu: bool = False):
+        self.store_agents_on_cpu = store_agents_on_cpu
+        self.storage_device = "cpu" if self.store_agents_on_cpu else device
+        self.agent = deepcopy(agent).to(self.storage_device)
+
+    def update_agent(self, agent):
+        self.agent = deepcopy(agent).to(self.storage_device)
+
+    def get_opponent(self):
+        # Always return the most recent agent
+        return self.agent
```

Comment:
```suggestion
    def get_opponent(self, agent_id):
        assert agent_id == 0, f"Self play only tracks the current agent. Expected agent id 0, got {agent_id}"
        # Always return the most recent agent
        return self.agent
```
I know that SelfPlay doesn't need this, but I'd like to have the same API for each curriculum. That way you can replace SelfPlay with FictitiousSelfPlay without making any code changes and have it work immediately. This will make more sense with my comments on FictitiousSelfPlay